### PR TITLE
fix broken links in docs, make docs builds stricter

### PR DESCRIPTION
**Pull request by jameslamb**
I noticed at https://rapidsai.github.io/legate-boost/README.html that some links to images and example READMEs are broken.

Screenshot 2024-07-19 at 11 14 30 AM

CI logs have some hints about those and other broken links:

```text
/__w/legate-boost/legate-boost/docs/source/README.md:34: WARNING: 'myst' cross-reference target not found: 'examples/probabalistic_regression/README.md' [myst.xref_missing]
/__w/legate-boost/legate-boost/docs/source/README.md:51: WARNING: 'myst' cross-reference target not found: 'examples/batch_training/README.md' [myst.xref_missing]
/__w/legate-boost/legate-boost/legateboost/legateboost.py:docstring of sklearn.utils._metadata_requests._MetadataRequester.get_metadata_routing:3: WARNING: undefined label: 'metadata_routing'
/__w/legate-boost/legate-boost/legateboost/legateboost.py:docstring of sklearn.utils._metadata_requests.RequestMethod.__get__.<locals>.func:3: WARNING: undefined label: 'metadata_routing'
```

([build link](https://github.com/rapidsai/legate-boost/actions/runs/10008731014/job/27666035153#step:11:47))

This proposes the following to fix them:

* copying the content of `examples/` into the docs source directory at build time (e.g. when running `make html`)
* adding those examples as **hidden** entries in the table of contents

And other fixes for the docs:

* removing `make.bat` file (I don't think anyone is building these docs on Windows, right?)
* running `sphinx-build --fail-on-warning`, so any warnings (like broken links) will result in loud CI failures
* small typo fixes

### How I tested this

```shell
conda env create \
    --name legate-boost-dev \
    --file ./conda/environments/all_cuda-118.yaml

source activate legate-boost-dev

# build + install the library
./build.sh

# build the docs
make -C docs clean html

# host the docs
python -m http.server \
    -d ./docs/build/html \
    1456
```

Clicked around the docs, saw all the images rendering successfully and links working.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -49,3 +50,8 @@
 # -- Options for extensions --------------------------------------------------
 
 autosummary_generate = True
+
+# ensure links to third-party docs work
+intersphinx_mapping = {
+    "sklearn": ("https://scikit-learn.org/stable/", None),
+}
```
Comment (jameslamb) [inline docs/source/conf.py:57]: This fixes warnings like this:

```text
docstring of sklearn.utils._metadata_requests.RequestMethod.__get__.<locals>.func:3: WARNING: undefined label: 'metadata_routing'
```

Context for this change:

* https://github.com/microsoft/LightGBM/issues/5954
* https://github.com/microsoft/LightGBM/pull/5956

File: README.md [hunk 2]
```diff
@@ -59,22 +59,18 @@ The following example shows a model combining linear and decision tree base lear
...
 ```
 
-<img src="examples/kernel_ridge_regression/kernel_ridge_regression.png" alt="drawing" width="400"/>
+<img src="_static/examples/kernel_ridge_regression.png" alt="drawing" width="400"/>
 
 ## Installation
 
-Dependencies:
-- cunumeric
-- sklearn
```
Comment (jameslamb) [inline README.md:76]: This list is out of date (omits `legate-boost`, `scipy`, and `numpy`). I think such lists in docs are hard to keep in sync with the project, and they don't provide much value beyond just directly inspecting `pyproject.toml`.

Proposing just removing this.

File: .github/workflows/github-actions.yml [hunk 3]
```diff
@@ -73,8 +73,7 @@ jobs:
       - name: Build legateboost docs
         working-directory: docs
         run: |
-          # invoke sphinx build using legate
-          legate --run-mode=python --module sphinx.cmd.build source build -b html
+          make html
```
Comment (jameslamb) [inline .github/workflows/github-actions.yml:76]: I wasn't sure why `legate` was being used to run the Sphinx build. I'm hoping it's ok to use the Sphinx `Makefile` instead, so we can customize the build in the way this PR proposes.

Comment (RAMitchell) [reply .github/workflows/github-actions.yml:76]: I remember not being able to get it to work the normal way, but if it works for you then that would be great :)

**Review by RAMitchell — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (RAMitchell) [reply README.md:76]: Agreed

File: .github/workflows/github-actions.yml (hunk 3 above)
Comment (jameslamb) [reply .github/workflows/github-actions.yml:76]: Yep it's been working!

Wonder if maybe having legate packages (#113) + building in a conda environment prevents whatever issues you were seeing before.

**Comment by jameslamb** [comment]
thanks!

**Pull request merged**